File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Is Graph Bipartite?/Java/bfs/Solution.java
import java.util.*;

class Solution {
    public boolean isBipartite(int[][] graph) {
        int[] colors = new int[graph.length];
        boolean[] visited = new boolean[graph.length];
        for (int start = 0; start < graph.length; start++) {
            if (visited[start]) {
                continue;
            }

            if (!bfs(graph, start, colors, visited)) {
                return false;
            }
        }

        return true;
    }

    private boolean bfs(int[][] graph, int start, int[] colors, boolean[] visited) {
        Deque<Integer> dq = new ArrayDeque<>();
        visited[start] = true;
        dq.offer(start);
        colors[start] = 1;

        while (!dq.isEmpty()) {
            int v = dq.poll();
            int vColor = colors[v];
            for (int next : graph[v]) {
                if (!visited[next]) {
                    visited[next] = true;
                    dq.offer(next);
                    colors[next] = 3 - vColor;
                    continue;
                }

                if (colors[next] == vColor) {
                    return false;
                }
            }
        }

        return true;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Is Graph Bipartite?/bfs/solution.py
from collections import deque

class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        visited = [False for _ in graph]
        node_color = [None for _ in graph]
        
        for start in range(len(visited)):
            if visited[start]:
                continue

            if not self.bfs(graph, visited, node_color, start):
                return False
        
        return True

    def bfs(self, graph, visited, node_color, start):
        dq = deque()
        dq.append([start, 'red'])
        visited[start] = True
        node_color[start] = 'red'

        while dq:
            current_node = dq.popleft()
            next_node_color = 'black' if current_node[1] == 'red' else 'red'
            for next in graph[current_node[0]]:
                if visited[next]:
                    if node_color[next] != next_node_color:
                        return False
                    else:
                        continue

                dq.append([next, next_node_color])
                visited[next] = True
                node_color[next] = next_node_color
        
        return True
